### PR TITLE
Protect storage pools that host volumes with a finalizer

### PR DESCRIPTION
**Pull request by blaisedias**
Modify the pool-operator to
1. add a finalizer to the pool CR if the number of
replicas in the pool is > 1
 only if the pool CR deletionTimestamp is not set.
2. remove the finalizer from the pool CR when the
number of replicas is 0.

### CODE REVIEW

**Comment by blaisedias** [comment]
bors try

**Review by jkryl — APPROVED** [review]

File: csi/moac/finalizer_helper.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+          log.debug(`added finalizer:${finalizer} to ${this.plural}:${instancename}`);
+        })
+        .catch((err:any) => {
+         log.error(`add finalizer:${finalizer} to ${this.plural}:${instancename}, update failed: code=${err.body.code}, reason=${err.body.reason}, ${err.body.message}`);
+        });
+  }
+
+  removeFinalizer(body: any, instancename: String, finalizer: String) {
+    if (body.metadata.finalizers == undefined) {
+        log.debug(`removeFinalizer(${instancename},${finalizer}), no finalizers on pool`);
```
Comment (paulyoong) [inline csi/moac/finalizer_helper.ts:60]: Should "pool" be specifically mentioned here? Everything else seems to be generic.

File: csi/moac/finalizer_helper.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+      this.k8sApi.getNamespacedCustomObject(
+          this.groupname,
+          this.version,
+          this.namespace,
+          this.plural,
+          instancename)
+          .then((customresource:any) => {
+              let body = customresource.body;
+              if (body.metadata.finalizers == undefined) {
+                  log.debug(`removeFinalizerFromCR(${instancename},${finalizer}), no finalizers on pool`);
```
Comment (paulyoong) [inline csi/moac/finalizer_helper.ts:143]: Again, should "pool" be mentioned here?

**Review by paulyoong — APPROVED** [review]

File: csi/moac/finalizer_helper.ts (hunk 2 above)
Comment (blaisedias) [reply csi/moac/finalizer_helper.ts:143]: yeah fixed now.

**Comment by blaisedias** [comment]
bors merge

**Review by paulyoong — COMMENTED** [review]
bors r+

**Comment by jkryl** [comment]
bors retry

**Pull request merged**